url8d3c4d17f5_catseye_JaC64_tgz-pJ8-JaC64J8/src/com/dreamfabric/jac64/SIDVoice6581.java

```diff
@@ -0,0 +1,740 @@
+    int tmp = ((memory[sidbase + 6] & 0xf0) + 8) << ADSR_BITS;
+    releaseFactor = decayExp[(memory[sidbase + 6] & 0x0f)];
+
+    if (DEBUG) {
+      System.out.println(getSIDNo() + "  Setting SR: " +
+			 Integer.toString(memory[sidbase + 6], 16) +
+			 " " + cycles + " diff " +
+			 (cycles - lastAttackTime) + " ADSRLv: " +
+			 (adsrLevel22 >> ADSR_BITS));
+    }
+    // Restart decay if sustain is reached, but we have lowered sustain level!!!
+    // ???
+
+    sr = data;
+    // If we are currently releasing...
+    if (adsrPhase == RELEASE) {
+      if (currentDecayFactor > releaseFactor) {
+	adsrBug = (int) (0x8000 * SAMPLES_PER_MICRO);
+	if (DEBUG) {
+	  System.out.println(getSIDNo() +
+			     "  REL_ADSR BUG Triggered for " + adsrBug +
+			     " samples" + " level: " +
+			     (adsrLevel22 >> ADSR_BITS));
+	}
+      }
+      currentDecayFactor = releaseFactor;
+    } else if (adsrPhase == SUSTAIN) {
+      if (tmp < adsrSusLevel22) {
+	currentDecayFactor = releaseFactor;
+	adsrPhase = RELEASE;
+      }
+    }
+    adsrSusLevel22 = tmp;
+  }
+
+
+  public void reset() {
+    init();
+    soundOff(0);
+  }
+
+  public void printStatus() {
+    System.out.println("SID: " + getSIDNo() + " ----------------------------");
+    System.out.println("Wave: " + WAVE[wave]);
+    System.out.println("Frequency: " + frq + "  PWid:" + pwid + " " + trueFrq);
+    System.out.println("ADSRLevel: " + (adsrLevel22 >> ADSR_BITS));
+    System.out.println("ADSRPhase: " + ADSR_PHASES[adsrPhase]);
+    System.out.println("AD reg: " + Integer.toString(memory[sidbase + 5], 16));
+    System.out.println("SR reg: " + Integer.toString(memory[sidbase + 6], 16));
+    System.out.println("WavePos: " + nextSample + " pulse => " + ((nextSample < pwid) ? "0xfff" : "0"));
+    if (ring)System.out.println("RING MODULATION");
+    if (sync)System.out.println("SYNCHRONIZATION");
+    if (testBit) System.out.println("TEST BIT SET!");
+  }
+
+
+  private int getSIDNo() {
+    return ((sidbase & 0xf) / 7);
+  }
+
+  private long lastCycles = 0;
+  private int decayTimes = 0;
+  // last sample
+  public int lastSample = 0;
+  // ADSR is 8 bits
+  public int adsrVol = 0;
+
+  long nanos;
+  long total;
+
+  int pwidArr[] = new int[SIDVoice6581.GENLEN];
+
+  // This is a crappy method... should not be used later...
+  public int[] generateSound(long cycles) {
+    updateSound(cycles);
+    // Cheat down generated to 0
+    generated = 0;
+    return outBuffer;
+  }
+
+  public void updatePulseWidth(long cycles) {
+    if ((wave & PULSE) != 0 && lastCycles > 0) {
+      int pos = (int) ((cycles - lastCycles) /
+		       (SIDVoice6581.CYCLES_PER_SAMPLE + 10));
+      pwidArr[pos % SIDVoice6581.GENLEN] =
+	((((memory[sidbase + 3]& 0xf) << 8) +
+	  memory[sidbase + 2]) * waveLen) / 4095;
+    } else {
+      pwid = ((((memory[sidbase + 3]& 0xf) << 8) +
+	       memory[sidbase + 2]) * waveLen) / 4095;
+    }
+  }
+
+  public void updateSound(long cycles) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

This method contains 179 lines of code, not including blank lines or lines with only comments, Java punctuation characters, identifiers, or literals. By comparison, 99% of the methods in the CodeGuru reference dataset contain fewer lines of code. Large methods might be difficult to read and have logic that is hard to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 472-726 into a separate method.